test/e2e/0xTrade.test.ts

```diff
@@ -0,0 +1,297 @@
+    // purposes of GPv2, these need to be either `address(0)` or the settlement
+    // contract.
+    takerAddress: ethers.constants.AddressZero,
+    senderAddress: ethers.constants.AddressZero,
+
+    // NOTE: Include no additional fees. I am not sure how this is used by
+    // market makers, but in theory this can be used to assign an additional
+    // fee, on top of the 0x protocol fee, to the GPv2 settlement contract.
+    feeRecipientAddress: ethers.constants.AddressZero,
+    makerFee: 0,
+    takerFee: 0,
+    makerFeeAssetData: "0x",
+    takerFeeAssetData: "0x",
+  };
+
+  const hash = ethers.utils._TypedDataEncoder.hash(
+    domain,
+    ZERO_EX_ORDER_TYPE_DESCRIPTOR,
+    order,
+  );
+
+  // NOTE: Use EIP-712 signing scheme for the order. The signature is just the
+  // ECDSA signature post-fixed with the signature scheme ID (0x02):
+  // <https://0x.org/docs/guides/v3-specification#signature-types>
+
+  const EIP712_SIGNATURE_ID = 0x02;
+  const { v, r, s } = ethers.utils.splitSignature(
+    await maker._signTypedData(domain, ZERO_EX_ORDER_TYPE_DESCRIPTOR, order),
+  );
+  const signature = ethers.utils.solidityPack(
+    ["uint8", "bytes32", "bytes32", "uint8"],
+    [v, r, s, EIP712_SIGNATURE_ID],
+  );
+
+  return { order, hash, signature };
+}
+
+describe("E2E: Can settle a 0x trade", () => {
+  let deployer: Wallet;
+  let solver: Wallet;
+  let trader: Wallet;
+  let marketMaker: Wallet;
+
+  let settlement: Contract;
+  let allowanceManager: Contract;
+  let domainSeparator: TypedDataDomain;
+
+  let owl: Contract;
+  let gno: Contract;
+
+  let zeroEx: {
+    exchange: Contract;
+    erc20Proxy: Contract;
+    domainSeparator: TypedDataDomain;
+  };
+
+  beforeEach(async () => {
+    const deployment = await deployTestContracts();
+
+    ({
+      deployer,
+      settlement,
+      allowanceManager,
+      wallets: [solver, trader, marketMaker],
+    } = deployment);
+
+    const { authenticator, owner } = deployment;
+    await authenticator.connect(owner).addSolver(solver.address);
+
+    const { chainId } = await ethers.provider.getNetwork();
+    domainSeparator = domain(chainId, settlement.address);
+
+    owl = await waffle.deployContract(deployer, ERC20, ["OWL", 18]);
+    gno = await waffle.deployContract(deployer, ERC20, ["GNO", 18]);
+
+    const exchange = await waffle.deployContract(
+      deployer,
+      Exchange.compilerOutput,
+      [chainId],
+    );
+    const erc20Proxy = await waffle.deployContract(
+      deployer,
+      ERC20Proxy.compilerOutput,
+    );
+
+    await erc20Proxy.addAuthorizedAddress(exchange.address);
+    await exchange.registerAssetProxy(erc20Proxy.address);
+
+    zeroEx = {
+      exchange,
+      erc20Proxy,
+      // NOTE: Domain separator parameters taken from:
+      // <https://0x.org/docs/guides/v3-specification#eip-712-usage>
+      domainSeparator: {
+        name: "0x Protocol",
+        version: "3.0.0",
```

Comment:
I am not sure I recall correctly, but did Lev say that they only integrated with v2?

Reply:
Yes, but I remember him saying that they plan on having v3 integration by the end of the month, so we decided to just integrate with v3 right off the bat.

Reply:
Are you sure they will support v3 (and not v4)? In telegram he said they skipped v3 because of the fee that 0x takes (this feature was waived in v4)

Reply:
Hmm, now I'm not sure anymore.

Reply:
If its v4, we can take advantage of [RFQ Orders](https://0xprotocol.readthedocs.io/en/latest/basics/orders.html#rfq-orders):

> RFQ orders are a stripped down version of standard limit orders, supporting fewer fields and a leaner settlement process